Add materialized view configuration for BigQuery (#4578)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/bigquery-configs.md

The `grant_access_to` config is not thread-safe when multiple views need to be authorized for the same dataset. The initial `dbt run` operation after a new `grant_access_to` config is added should therefore be executed in a single thread. Subsequent runs using the same configuration will not attempt to re-apply existing access grants, and can make use of multiple threads.

<VersionBlock firstVersion="1.7">

## Materialized views

The BigQuery adapter supports [materialized views](https://cloud.google.com/bigquery/docs/materialized-views-intro)
with the following configuration parameters:

| Parameter                    | Type                 | Required | Default   | Change Monitoring Support | Reference                                        |
|------------------------------|----------------------|----------|-----------|---------------------------|--------------------------------------------------|

> | `on_configuration_change`    | STRING               | NO       | `'apply'` | N/A                       |                                                  |

Reviewer: Without quotes should be easier to read and use since quotes are not required in YAML configuration.
```suggestion
| `on_configuration_change`    | STRING               | NO       | `apply` | N/A                       |                                                  |
```

| `cluster_by`                 | LIST[STRING]         | NO       |           | DROP/CREATE               | [Clustering](#clustering-clause)                 |
| `partition_by`               | DICT                 | NO       |           | DROP/CREATE               | [Partitioning](#partition-clause)                |

> | `enable_refresh`             | BOOLEAN              | NO       | `True`    | ALTER                     | [Auto-refresh](#auto-refresh)                    |

Reviewer: Lowercase `true` is valid in both Jinja and YAML (and was even required to be lowercase in Jinja at same point).

```suggestion
| `enable_refresh`             | BOOLEAN              | NO       | `true`    | ALTER                     | [Auto-refresh](#auto-refresh)                    |
```

| `refresh_interval_minutes`   | FLOAT                | NO       | `30`      | ALTER                     | [Auto-refresh](#auto-refresh)                    |
| `max_staleness` (in Preview) | INTERVAL             | NO       |           | ALTER                     | [Auto-refresh](#auto-refresh)                    |
| `description`                | STRING               | NO       |           | ALTER                     |                                                  |
| `labels`                     | DICT[STRING, STRING] | NO       |           | ALTER                     | [Labels](#specifying-labels)                     |
| `hours_to_expiration`        | INTEGER              | NO       |           | ALTER                     | [Table expiration](controlling-table-expiration) |

> | `kms_key_name`               | STRING               | NO       |           | ALTER                     | [KMS encryption](#using-kms-encryption)          |

Reviewer: The table can be one column thinner if doing a hyperlink like [`kms_key_name`](#using-kms-encryption) in the first column instead.

Author: Done

    on_configuration_change="apply" | "continue" | "fail",
    [cluster_by](#clustering-clause)="<field-name>" | ["<field-name>"],
    [partition_by](#partition-clause)={
        "field": "<field-name>",
        "data_type": "timestamp" | "date" | "datetime" | "int64",

        # only if `data_type` is not 'int64'
        "granularity": "hour" | "day" | "month" | "year,

        # only if `data_type` is 'int64'
        "range": {
            "start": <integer>,
            "end": <integer>,
            "interval": <integer>,
        }
    },

    # auto-refresh options
    [enable_refresh](#auto-refresh)= true | false,
    [refresh_interval_minutes](#auto-refresh)=<float>,
    [max_staleness](#auto-refresh)="<interval>",

    # additional options
    [description](/reference/resource-properties/description)="<description>",
    [labels](#specifying-labels)={
        "<label-name>": "<label-value>",
    },
    [hours_to_expiration](#acontrolling-table-expiration)=<integer>,
    [kms_key_name](##using-kms-encryption)="<path_to_key>",
) }}
```

</File>

</TabItem>

</Tabs>

Many of these parameters correspond to their table counterparts and have been linked above.

> The set of parameters which are unique to materialized views covers auto-refresh functionality, which is covered [below](#auto-refresh).

Reviewer 2: ```suggestion
The set of parameters unique to materialized views covers [auto-refresh functionality](#auto-refresh).
```

Find more information about these parameters in the BigQuery docs:
- [CREATE MATERIALIZED VIEW statement](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#create_materialized_view_statement)
- [materialized_view_option_list](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#materialized_view_option_list)

### Auto-refresh

| Parameter                    | Type         | Required | Default | Change Monitoring Support |
|------------------------------|--------------|----------|---------|---------------------------|
| `enable_refresh`             | `<boolean>`  | no       | `true`  | alter                     |
| `refresh_interval_minutes`   | `<float>`    | no       | `30`    | alter                     |
| `max_staleness` (in Preview) | `<interval>` | no       | `none`  | alter                     |

BigQuery supports [automatic refresh](https://cloud.google.com/bigquery/docs/materialized-views-manage#automatic_refresh) configuration for materialized views.

> By default, a materialized view will automatically refresh within 5 minutes of changes in the base table, but no more frequently than every 30 minutes.

Reviewer 2: ```suggestion
By default, a materialized view will automatically refresh within 5 minutes of changes in the base table, but not more frequently than once every 30 minutes.
```

> BigQuery only officially supports configuration of the frequency (the "30 minutes" part);

Reviewer 2: ```suggestion
BigQuery only officially supports the configuration of the frequency (the "once every 30 minutes" frequency);
```

> however, there is a feature in preview that allows for configuration of the staleness (the "5 minutes" part).

Reviewer 2: ```suggestion
however, there is a feature in preview that allows for the configuration of the staleness (the "5 minutes" refresh).
```

dbt will monitor these parameters for changes and apply them using an `ALTER` statement.

Find more information about these parameters in the BigQuery docs:
- [materialized_view_option_list](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#materialized_view_option_list)
- [max_staleness](https://cloud.google.com/bigquery/docs/materialized-views-create#max_staleness)

### Limitations

As with most data platforms, there are limitations associated with materialized views. Some worth noting include:

> - Materialized view SQL has a [limited feature set](https://cloud.google.com/bigquery/docs/materialized-views-create#supported-mvs)

Reviewer 2: ```suggestion
- Materialized view SQL has a [limited feature set](https://cloud.google.com/bigquery/docs/materialized-views-create#supported-mvs).
```

> - Materialized view SQL cannot be updated; the materialized view must go through a `--full-refresh` (DROP/CREATE)

Reviewer 2: ```suggestion
- Materialized view SQL cannot be updated; the materialized view must go through a `--full-refresh` (DROP/CREATE).
```

> - The `partition_by` clause on a materialized view must match that of the underlying base table

Reviewer 2: ```suggestion
- The `partition_by` clause on a materialized view must match that of the underlying base table.
```

> - While materialized views can have descriptions, materialized view *columns* cannot

Reviewer 2: ```suggestion
- While materialized views can have descriptions, materialized view *columns* cannot.
```

> - Recreating/dropping the base table requires recreating/dropping the materialized view

Reviewer 2: ```suggestion
- Recreating/dropping the base table requires recreating/dropping the materialized view.
```

| [`refresh_interval_minutes`](#auto-refresh)            | `<float>`              | no       | `30`    | alter                     |
| [`max_staleness`](#auto-refresh) (in Preview)          | `<interval>`           | no       | `none`  | alter                     |
| `description`                                          | `<string>`             | no       | `none`  | alter                     |
| [`labels`](#specifying-labels)                         | `{<string>: <string>}` | no       | `none`  | alter                     |
| [`hours_to_expiration`](#controlling-table-expiration) | `<integer>`            | no       | `none`  | alter                     |
| [`kms_key_name`](#using-kms-encryption)                | `<string>`             | no       | `none`  | alter                     |

<Tabs
  groupId="config-languages"
  defaultValue="project-yaml"
  values={[
    { label: 'Project file', value: 'project-yaml', },
    { label: 'Property file', value: 'property-yaml', },
    { label: 'Config block', value: 'config', },
  ]
}>

<TabItem value="project-yaml">

<File name='dbt_project.yml'>

> directly to the materialized view in place.

Reviewer: Maybe this can just be deleted altogether?

```suggestion
```

It looks like a rogue line that is causing some funky rendering:

<img width="350" alt="image" src="https://github.com/dbt-labs/docs.getdbt.com/assets/44704949/69af75fa-be7d-465e-b2ad-f29c48a200b5">
